### PR TITLE
Bool method

### PR DESCRIPTION
**Pull request by 0lch4**

### CODE REVIEW

File: src/math-land/bool.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    if not que:
+        print(bool(que))
+
+    que.put("person")
+
+    if que:
+        print(bool(que))
+
+    print(bool(PlainClass()))
```
Comment (CircuitSacul) [inline src/math-land/bool.py:37]: I feel like a more readable example would be something like this:

```py
que = Queue()

if que:
    print("que is non-empty")
else:
    print("que is empty")

que.put("person")

if que:
    print("que is non-empty")
else:
    print("que is empty")

# you can also do this
print("bool(que) =", bool(que))

# class without a `__bool__` method are always Truthy
print("bool(PlainClass())=", bool(PlainClass()))
```

File: src/math-land/bool.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+from collections.abc import Sequence
+from typing import Generic, TypeVar
+
+T = TypeVar("T")
+
+
+class Queue(Generic[T]):
+    def __init__(self, items: Sequence[T] | None = None) -> None:
```
Comment (CircuitSacul) [inline src/math-land/bool.py:8]: Just a thought - does having advanced typing features distract from the main point (the dunder methods)?

cc @trag1c @jamesmurphy-mc

**Review by CircuitSacul — CHANGES_REQUESTED** [review]

File: src/math-land/bool.py (hunk 1 above)
Comment (trag1c) [reply src/math-land/bool.py:37]: Quick reminder that `print(f"bool(que) =", bool(que))` can be replaced with just `print(f"{bool(que) = }")`

Comment (CircuitSacul) [reply src/math-land/bool.py:37]: ah right

File: src/math-land/bool.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    que.put("person")
+
+    if que:
+        print("que is non-empty")
+    else:
+        print("que is empty")
+
+    print(f"{bool(que)=}") #True
+
+    # class without a `__bool__` method are always Truthy
```
Comment (trag1c) [inline src/math-land/bool.py:44]:
```suggestion
    # instances of classes without a `__bool__` method are always truthy
```

**Review by trag1c — CHANGES_REQUESTED** [review]

**Review by trag1c — APPROVED** [review]

File: src/math-land/bool.py (hunk 2 above)
Comment (trag1c) [reply src/math-land/bool.py:8]: I personally don't mind it at all

**Review by CircuitSacul — APPROVED** [review]

**Comment by trag1c** [comment]
Make black happy :+1:

File: src/math-land/bool.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        self.items = list(items or [])
+
+    def __bool__(self) -> bool:
+        return len(self.items) > 0
+
+    def put(self, item: T) -> None:
+        self.items.append(item)
+
+    def get(self) -> T:
+        return self.pop(0)
```
Comment (James-Ansley) [inline src/math-land/bool.py:18]: Should be `return self.items.pop(0)`?

**Review by James-Ansley — CHANGES_REQUESTED** [review]

**Comment by James-Ansley** [comment]
How do people feel about possibly duplicating some dunders? I feel like this could be a really nice example showing that `__len__` is called if `__bool__` is not implemented – but I can see that `__len__` is already assigned.

**Comment by trag1c** [comment]
> How do people feel about possibly duplicating some dunders? I feel like this could be a really nice example showing that `__len__` is called if `__bool__` is not implemented – but I can see that `__len__` is already assigned.

Feels inevitable

**Comment by bswck** [comment]
> How do people feel about possibly duplicating some dunders? I feel like this could be a really nice example showing that `__len__` is called if `__bool__` is not implemented – but I can see that `__len__` is already assigned.

Very good idea IMO!

**Review by bswck — APPROVED** [review]

**Pull request merged**